android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt

```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+private fun updateNotification(
+    context: Context, fcmMessage: MessageFcmMessage) {
+    val user = Person.Builder().setName("You").build()
+    val sender = Person.Builder()
+        .setName(fcmMessage.sender.fullName)
+        .setIcon(IconCompat.createWithBitmap(fetchBitmap(fcmMessage.sender.avatarURL)))
+        .build()
+
+    var isGroupConversation = true
+    val title = when (fcmMessage.recipient) {
+        is Recipient.Stream -> "#${fcmMessage.recipient.stream} > ${fcmMessage.recipient.topic}"
+        // TODO use proper title for GroupPM, we will need
+        // to have a way to get names of PM users here.
+        is Recipient.GroupPm -> context.getString(R.string.group_pm, fcmMessage.recipient.pmUsers)
+        is Recipient.Pm -> {
+            isGroupConversation = false
+            fcmMessage.sender.fullName
+        }
+    }
+    val groupKey = "${fcmMessage.identity?.realmUri.toString()};${fcmMessage.identity?.userId}"
+    val conversationKey = "$groupKey;$title"
+    val extraData = Bundle()
+    extraData.putString("conversationKey", conversationKey)
+    var (notificationId, notification) = getActiveNotification(context, conversationKey)
+    notificationId = notificationId?: fcmMessage.zulipMessageId
+
+    val messagingStyle = when (notification) {
+        null -> NotificationCompat.MessagingStyle(user)
+        else -> NotificationCompat.MessagingStyle
+            .extractMessagingStyleFromNotification(notification)
+            ?: NotificationCompat.MessagingStyle(user)
+    }
+    messagingStyle
+        .setConversationTitle(title)
+        .setGroupConversation(isGroupConversation)
+        .addMessage(fcmMessage.content, fcmMessage.timeMs, sender)
+
+    val messageCount = messagingStyle.messages.size
+
+    val builder = NotificationCompat.Builder(context, CHANNEL_ID)
+    builder.apply {
+        color = Color.rgb(100, 146, 254)
+        if (BuildConfig.DEBUG) {
+            setSmallIcon(R.mipmap.ic_launcher)
+        } else {
+            setSmallIcon(R.drawable.zulip_notification)
+        }
+        setAutoCancel(true)
+        setStyle(messagingStyle)
+        setGroup(groupKey)
+        setContentIntent(createViewPendingIntent(fcmMessage, context))
+        setNumber(messageCount)
+        extras = extraData
+    }
 
-    val dismissIntent = Intent(context, NotificationIntentService::class.java)
-    dismissIntent.action = ACTION_CLEAR
-    val piDismiss = PendingIntent.getService(context, 0, dismissIntent, 0)
-    val action = Notification.Action(android.R.drawable.ic_menu_close_clear_cancel, "Clear", piDismiss)
-    builder.addAction(action)
+    val summaryNotification = createSummaryNotification(context, fcmMessage, groupKey)
 
-    val soundUri = getNotificationSoundUri(context)
-    val audioAttr = AudioAttributes.Builder()
-        .setUsage(AudioAttributes.USAGE_NOTIFICATION).build()
-    builder.setSound(soundUri, audioAttr)
```

Comment:
This is one chunk that I notice doesn't have a counterpart in the new code.

Looking at the API docs, I see that it has no effect on Android 8 O and up!
https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#setSound(android.net.Uri)
So it wasn't doing us a lot of good. But: that's because we're supposed to be calling the corresponding method on the notification *channel*'s builder instead:
https://developer.android.com/reference/androidx/core/app/NotificationChannelCompat.Builder#setSound(android.net.Uri,%20android.media.AudioAttributes)

So it seems like that's something we neglected to take care of when we started using a notification channel a few years ago. But seeing this code and deleting it gives us a nice reminder of how we should be doing it. Would you add a few lines to our `createNotificationChannel` to fix that?

The very cleanest thing would be to make sure we continue to do it on the notification itself, for the sake of pre-Android 8 devices, as well as on the channel. But at this point those devices are well into the single-digit percent (they [were 9.0%](https://chat.zulip.org/#narrow/stream/48-mobile/topic/platform.20versions/near/1012192) almost a year ago), so graceful degradation is okay, and I'd be happy with a version that only does it on the channel.

Reply:
I have added `audioAttr` use in channel as defined previously for notification builder.
`NotificationCompat.Builder` doesn't allow me to call `setSound` with `audioAttr` (there is no function with that signature in `NotificationCompat.Builder`) so i don't think i can easily add it there.